doc/designs/fixed-size-large-numbers.md

```diff
@@ -0,0 +1,711 @@
+
+    if (r != NULL) {
+        r->data = a;
+        r->flags |= (a->is_dynamically_allocated ? BN_FLG_MALLOCED);
+        r->flags |= (a->is_securely_allocated ? BN_FLG_SECURE);
+        a->is_acquired = 1;
+    }
+
+    return r;
+}
+```
+
+This should be avoided, but we retain this possibility just in case its use
+is unavoidable.
+
+Mutability
+----------
+
+[Mutability]: #mutability
+
+The understanding is that within a `OSSL_FN` API, `dsize` is immutable as
+soon as a `OSSL_FN` has been allocated to its target size, except for when
+the `OSSL_FN` instance is freed. However, when accessed through the `BIGNUM`
+type (i.e. by the `BIGNUM` API), the `OSSL_FN` size may be reallocated to
+allow a larger size than initially allocated, and `dsize` may be modified
+accordingly.
+
+The `OSSL_FN_CTX`  API is much more strict.  The size of an `OSSL_FN_CTX`
+instance is immutable after it has been allocated, except when it is freed.
+
+In the case where an `OSSL_FN` is "acquired" (i.e. the calling code holds a
+pointer to a `OSSL_FN` and to a `BIGNUM` that wraps it), `BIGNUM` must
+consider the `OSSL_FN` size as immutable.
+
+Dropping `top` from `BIGNUM`
+----------------------------
+
+[Dropping `top` from `BIGNUM`]: #dropping-top-from-bignum
+
+With this design, the `top` field is dropped, meaning that a `BIGNUM` will
+no longer be able to diminish its memory footprint. The reason is that the
+`OSSL_FN` functionality would need to update it, but that runs counter to
+the ideas behind the `OSSL_FN` functionality in terms of mutability.
+
+This may come at a performance cost for users calling the `BIGNUM`
+API. However, it's our judgement that this is acceptable to get better
+overall safety.
+
+Memory functionality for `OSSL_FN`
+----------------------------------
+
+[Memory functionality for `OSSL_FN`]: #memory-functionality-for-ossl_fn
+
+We anticipate that we will need the following functions to allocate and
+deallocate `OSSL_FN`s:
+
+```c
+OSSL_FN *OSSL_FN_new(size_t size);
+void OSSL_FN_free(OSSL_FN *f);
+```
+
+Memory functionality for `OSSL_FN_CTX`
+--------------------------------------
+
+[Memory functionality for `OSSL_FN_CTX`]: #memory-functionality-for-ossl_fn_ctx
+
+We anticipate that the `OSSL_FN_CTX` API will look very much like the
+`BN_CTX` API, except for the allocation functionality:
+
+```c
+OSSL_FN_CTX *OSSL_FN_CTX_new(OSSL_LIB_CTX *libctx, size_t arena_size);
+OSSL_FN_CTX *OSSL_FN_CTX_secure_new(OSSL_LIB_CTX *libctx, size_t arena_size);
+void OSSL_FN_CTX_free(OSSL_FN_CTX *ctx);
+
+OSSL_FN *OSSL_FN_CTX_get(OSSL_FN_CTX *ctx, size_t size);
+```
+
+*Something to be noted is that `OSSL_FN_CTX_secure_new()` allocates the
+whole arena in secure memory.  The impact compared to allocating individual
+`OSSL_FN` instances in secure memory is considered minimal.*
+
+If the variant of `OSSL_FN_CTX` *with frames* is chosen, the following
+functions will also have to be defined:
+
+```c
+void OSSL_FN_CTX_start(OSSL_FN_CTX *ctx);
```

Comment:
In theory OSSL_FN_CTX_start() can fail if you exhaust the arena by calling start() repeatedly. Cannot it?